apollo-normalized-cache/src/commonMain/kotlin/com/apollographql/apollo/cache/normalized/MemoryCache.kt

```diff
@@ -50,40 +52,53 @@ class MemoryCache(
   }
 
   override fun remove(cacheKey: CacheKey, cascade: Boolean): Boolean {
-    val cachedEntry = lruCache.remove(cacheKey.key)
-    if (cascade && cachedEntry != null) {
-      for (cacheReference in cachedEntry.record.referencedFields()) {
+    val cacheEntry = lruCache.remove(cacheKey.key)
+
+    if (cascade && cacheEntry != null) {
+      for (cacheReference in cacheEntry.record.referencedFields()) {
         remove(CacheKey(cacheReference.key), true)
       }
     }
 
-    val removeFromNextCacheResult = nextCache?.remove(cacheKey, cascade) ?: false
-
-    return cachedEntry != null || removeFromNextCacheResult
+    return cacheEntry != null || nextCache?.remove(cacheKey, cascade) ?: false
   }
 
-  override fun performMerge(apolloRecord: Record, oldRecord: Record?, cacheHeaders: CacheHeaders): Set<String> {
-    return if (oldRecord == null) {
-      lruCache[apolloRecord.key] = CacheEntry(
-          record = apolloRecord,
+  override fun merge(record: Record, cacheHeaders: CacheHeaders): Set<String> {
+    if (cacheHeaders.hasHeader(ApolloCacheHeaders.DO_NOT_STORE)) {
+      return emptySet()
+    }
+
+    val oldRecord = loadRecord(record.key, cacheHeaders)
+    val changedKeys = if (oldRecord == null) {
+      lruCache[record.key] = CacheEntry(
+          record = record,
           expireAfterMillis = expireAfterMillis
       )
-      apolloRecord.keys()
+      record.keys()
     } else {
-      oldRecord.mergeWith(apolloRecord).also {
+      oldRecord.mergeWith(record).also {
         //re-insert to trigger new weight calculation
-        lruCache[apolloRecord.key] = CacheEntry(
+        lruCache[record.key] = CacheEntry(
             record = oldRecord,
             expireAfterMillis = expireAfterMillis
         )
       }
     }
+
+    return changedKeys + nextCache?.merge(record, cacheHeaders).orEmpty()
+  }
+
+  override fun merge(records: Collection<Record>, cacheHeaders: CacheHeaders): Set<String> {
+    if (cacheHeaders.hasHeader(ApolloCacheHeaders.DO_NOT_STORE)) {
+      return emptySet()
+    }
+    return records.flatMap { record -> merge(record, cacheHeaders) }.toSet()
```

Comment:
I was going to write above that maybe we should keep the common `merge(collection)` in the parent class but given the implementation is so small, it doesn't make sense indeed. 👍